### ISSUE TITLE
Include packages for x86/x64 onigwrap.dll, and dynamic binding between them

### ISSUE THREAD

**Issue by dmitry-medvedev** [danipen/TextMateSharp#51]
We use TextMateSharp in our projects; on our tests, projects compiled with AnyCPU platform either do not copy onigwrap.dll to the bin folder (in case of packages.config), or copy x86 version of onigrwrap.dll (in case of sdk-style projects/package references). On the x64 platform, the project cannot load this dll and fails to run.

Please consider providing additional packages like TextMateShar.onigwrap.dll.Native.win-x64/TextMateShar.onigwrap.dll.Native.win-x86,

and implementing run-time binding to the correct version of onigwrap.dll depending on the platform.

Please take a look at ClearScript project which does it this way:
https://www.nuget.org/packages/Microsoft.ClearScript

(different platforms are handled inside V8SplitProxyNative.Generated.cs)

**Comment by danipen** [comment]
@dmitry-medvedev are you planning a PR for this or do you want me to take a look (not sure if I will able to take a look soon since I quite busy now).

**Comment by danipen** [comment]
Ok than. Thanks.

**Comment by danipen** [comment]
Not sure I understand the proposed solution. Are you proposing separating different nuget packages for each platform? In that case, how would you need to reference each nuget reference in the .csproj based on the current platform?

Note that, right now, the nuget package packs every version of the native libraries, and it places it on the right path, so the runtime is able to load them based on the platform:
* `runtimes/win-x86/native/onigwrap.dll` for [x86](https://github.com/danipen/TextMateSharp/blob/7066e69c21b7a416c150abfb4b6bbcd1c7cf97b2/src/TextMateSharp/TextMateSharp.csproj#L45)
* `runtimes/win-x64/native/onigwrap.dll` for [x64](https://github.com/danipen/TextMateSharp/blob/7066e69c21b7a416c150abfb4b6bbcd1c7cf97b2/src/TextMateSharp/TextMateSharp.csproj#L50)
* `runtimes/linux/native/libonigwrap.so` for [linux](https://github.com/danipen/TextMateSharp/blob/7066e69c21b7a416c150abfb4b6bbcd1c7cf97b2/src/TextMateSharp/TextMateSharp.csproj#L57)
* `runtimes/osx/native/libonigwrap.dylib` for [macOS](https://github.com/danipen/TextMateSharp/blob/7066e69c21b7a416c150abfb4b6bbcd1c7cf97b2/src/TextMateSharp/TextMateSharp.csproj#L63)

An existing issue is that, in debug mode, using Visual Studio, [only the x86 version is copied](https://github.com/danipen/TextMateSharp/blob/7066e69c21b7a416c150abfb4b6bbcd1c7cf97b2/src/TextMateSharp/TextMateSharp.csproj#L25) to the output dir, and probably, using x64 won't work (in my case it works fine for the Demo project but it fails for the unit tests).

**Comment by danipen** [comment]
So probably, if we find a better way of [copying the native deps](https://github.com/danipen/TextMateSharp/blob/7066e69c21b7a416c150abfb4b6bbcd1c7cf97b2/src/TextMateSharp/TextMateSharp.csproj#LL23C1-L39C15) into the `bin\Debug\netstandard2.0` output-dir, we may fix the issue.

**Comment by dmitry-medvedev** [comment]
Yes, separate NuGet packages for every platform looks like an overkill. We will try the solution you proposed.

**Comment by danipen** [comment]
@dmitry-medvedev I have some time to have a look. Could you please confirm if #52 resolves the problem you were facing?

**Comment by dmitry-medvedev** [comment]
Many thanks for that!

We will test changes in #52 and will let you know ASAP. 
Changes to the theme names work great for us.

Dmitry

**Comment by danipen** [comment]
Awesome. Thanks for the contribution BTW! 😉

**Comment by dmitry-medvedev** [comment]
We've tested the new interop code with different platforms, and it works perfectly. Thank you!

**Comment by danipen** [comment]
A new version of TextMateSharp is available in NuGet including the PR that fixes this issue: https://www.nuget.org/packages/TextMateSharp/1.0.55

**Comment by danipen** [comment]
BTW and out of  curiosity ... are you using AvaloniaEdit for the editor or are you using a different implementation?

**Comment by dmitry-medvedev** [comment]
No, we're not using Avalonia, it's our own implementation.

**Comment by dmitry-medvedev** [comment]
Hi Daniel,

I thought you might be interested. We've released a new version of AlterNET Studio, with a new TextMate-based parser as one of the major features:
https://www.alternetsoft.com/news/alternet-studio-9-0-released

There are a couple of things that we had to work around when using the TextMateSharp package. Maybe that's something you may address in future releases:

1. TextMateSharp depends on System.Text.Json of version 7.0.0.2, while other libraries that we use, such as Microsoft Code Analysis, use version 7.0.0.0. I wonder if there is a specific reason why you're using that specific version or if it can be downgraded to 7.0.0.0.

2. We found that for .net applications using .Net Framework 4.6.2, x86 and x64 onigwrap libraries are not copied in the bin folder. On our tests, either the x64 or x86 bit version is copied. We solved this issue by adding onigwrap libraries to our own package that uses TextMateSharp with the following build target:

<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
  <ItemGroup Condition="'$(TargetFramework)' == '' Or '$(TargetFramework.TrimEnd(`0123456789`))' == 'net'">
    <None Include="$(MSBuildThisFileDirectory)..\runtimes\win-x64\native\onigwrap-x64.dll">
      <Link>onigwrap-x64.dll</Link>
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
      <Visible>false</Visible>
    </None>
    <None Include="$(MSBuildThisFileDirectory)..\runtimes\win-x86\native\onigwrap-x86.dll">
      <Link>onigwrap-x86.dll</Link>
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
      <Visible>false</Visible>
    </None>
  </ItemGroup>
</Project>

Kind regards,
Dmitry

**Comment by danipen** [comment]
> I wonder if there is a specific reason why you're using that specific version or if it can be downgraded to 7.0.0.0.

No prob about downgrading. I was just using latest.

**Comment by danipen** [comment]
> We found that for .net applications using .Net Framework 4.6.2, x86 and x64 onigwrap libraries are not copied in the bin folder. On our tests, either the x64 or x86 bit version is copied.

Not sure about this one. The nugget package should deploy the native deps automatically...

Are you consuming TexmateSharp from the nuget package?

**Comment by dmitry-medvedev** [comment]
[WindowsFormsApp7.zip](https://github.com/danipen/TextMateSharp/files/11651712/WindowsFormsApp7.zip)

I'm attaching a sample project created with .NET Framework referencing TextMateSharp and TextMateSharp.Grammars packages. On our tests it does not copy onigwrap DLLs to the bin folder.

**Comment by danipen** [comment]
> No prob about downgrading. I was just using latest.

Same for this one. Add a PR if you want and I'll be happy to merge it 😊. Thanks!
